File: 20-actors/kawaraban/methods/test_ingest.py
#!/usr/bin/env python3
"""kawaraban — tests for the offline outlet normalizer membrane (ingest.py)."""
from __future__ import annotations
import json
import pathlib

from ingest import normalize_batch, normalize_record, live_allowed, IngestRefused

BATCH = pathlib.Path(__file__).resolve().parent.parent / "data" / "ingest" / "sample-batch.json"


def _records():
    return json.loads(BATCH.read_text(encoding="utf-8"))


def test_batch_accepts_clean_refuses_violations():
    ok, refused = normalize_batch(_records())
    assert len(ok) == 2, [a[":news.article/id"] for a in ok]
    assert len(refused) == 3, refused
    blob = " ".join(refused)
    assert "G4" in blob and "G1" in blob  # body/paywall (G4) + verdict (G1)


def test_refuses_full_body():
    try:
        normalize_record({"outlet": "o", "url": "u", "body": "the whole thing"})
        assert False, "expected G4 refusal"
    except IngestRefused as e:
        assert "G4" in str(e)


def test_refuses_paywall():
    try:
        normalize_record({"outlet": "o", "url": "u", "access": "paywall"})
        assert False, "expected G4 refusal"
    except IngestRefused as e:
        assert "G4" in str(e)


def test_refuses_verdict():
    try:
        normalize_record({"outlet": "o", "url": "u", "verdict": True})
        assert False, "expected G1 refusal"
    except IngestRefused as e:
        assert "G1" in str(e)


def test_requires_url():
    try:
        normalize_record({"outlet": "o", "headline": "h"})
        assert False, "expected G4/G5 url refusal"
    except IngestRefused as e:
        assert "url" in str(e).lower()


def test_excerpt_truncated_to_280():
    rec = normalize_record({"outlet": "o", "url": "u", "excerpt": "x" * 500})
    assert len(rec[":news.article/excerpt"]) == 280
    assert rec["_excerpt_truncated"] is True


def test_normalized_is_mirror_kind():
    rec = normalize_record({"outlet": "outlet.nhk", "url": "https://x", "headline": "h", "asOf": 5})
    assert rec[":news.article/kind"] == ":mirror"
    assert rec[":news.article/sourcing"] == ":representative"


def test_live_refused_at_r0():
    assert live_allowed() is False  # no operator gate env set


if __name__ == "__main__":
    import sys
    tests = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    failed = 0
    for t in tests:
        try:
            t()
        except Exception as e:  # noqa: BLE001
            failed += 1
            print(f"FAIL {t.__name__}: {e}")
    print(f"ingest: {len(tests) - failed}/{len(tests)} passed")
    sys.exit(1 if failed else 0)

File: 20-actors/kawaraban/methods/test_ingest.clj
(ns kawaraban.methods.test-ingest
  "kawaraban — tests for the offline outlet normalizer membrane (ingest.clj). 1:1 port of test_ingest.py."
  (:require [clojure.test :refer [deftest is]]
            [clojure.string :as str]
            [clojure.java.io :as io]
            [cheshire.core :as json]
            [kawaraban.methods.ingest :as ingest]))

(defn- sample-batch-path []
  (or (some-> (io/resource "kawaraban/data/ingest/sample-batch.json") str)
      "20-actors/kawaraban/data/ingest/sample-batch.json"))

(defn- load-records []
  (json/parse-stream (io/reader (sample-batch-path)) false))

(defn- refused? [thunk]
  (try
    (thunk)
    false
    (catch clojure.lang.ExceptionInfo e
      (= :kawaraban/ingest-refused (-> e ex-data :kawaraban/error)))))

(defn- refusal-message [thunk]
  (try
    (thunk)
    nil
    (catch clojure.lang.ExceptionInfo e
      (.getMessage e))))

(deftest test-batch-accepts-clean-refuses-violations
  (let [[ok refused] (ingest/normalize-batch (load-records))]
    (is (= 2 (count ok)) ok)
    (is (= 3 (count refused)) refused)
    (let [blob (str/join " " refused)]
      (is (str/includes? blob "G4") blob)
      (is (str/includes? blob "G1") blob))))

(deftest test-refuses-full-body
  (is (refused? #(ingest/normalize-record {"outlet" "o" "url" "u" "body" "the whole thing"})))
  (is (str/includes? (or (refusal-message #(ingest/normalize-record {"outlet" "o" "url" "u" "body" "the whole thing"})) "") "G4")))

(deftest test-refuses-paywall
  (is (refused? #(ingest/normalize-record {"outlet" "o" "url" "u" "access" "paywall"})))
  (is (str/includes? (or (refusal-message #(ingest/normalize-record {"outlet" "o" "url" "u" "access" "paywall"})) "") "G4")))

(deftest test-refuses-verdict
  (is (refused? #(ingest/normalize-record {"outlet" "o" "url" "u" "verdict" true})))
  (is (str/includes? (or (refusal-message #(ingest/normalize-record {"outlet" "o" "url" "u" "verdict" true})) "") "G1")))

(deftest test-requires-url
  (is (refused? #(ingest/normalize-record {"outlet" "o" "headline" "h"})))
  (is (str/includes? (str/lower-case (or (refusal-message #(ingest/normalize-record {"outlet" "o" "headline" "h"})) "")) "url")))

(deftest test-excerpt-truncated-to-280
  (let [rec (ingest/normalize-record {"outlet" "o" "url" "u" "excerpt" (apply str (repeat 500 "x"))})]
    (is (= 280 (count (get rec ":news.article/excerpt"))))
    (is (true? (get rec "_excerpt_truncated")))))

(deftest test-normalized-is-mirror-kind
  (let [rec (ingest/normalize-record {"outlet" "outlet.nhk" "url" "https://x" "headline" "h" "asOf" 5})]
    (is (= ":mirror" (get rec ":news.article/kind")))
    (is (= ":representative" (get rec ":news.article/sourcing")))))

(deftest test-live-refused-at-r0
  (is (false? (ingest/live-allowed))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (clojure.test/run-tests (quote kawaraban.methods.test-ingest))]
    (System/exit (if (zero? (+ fail error)) 0 1))))
